### PR TITLE
isleaf is isempty

### LINKED ISSUE #52: empty namedtuple is not a leaf
```julia
julia> Functors.isleaf((;))
false
```
I guess we should consider this a bug.

### PR DESCRIPTION
**Pull request by CarloLucibello**
implement the suggestion in https://github.com/FluxML/Functors.jl/issues/52#issuecomment-1327734936

Fix #52

### CODE REVIEW

File: test/basics.jl [hunk 1]
```diff
@@ -149,6 +149,14 @@ end
   @test_throws Exception functor(NamedTuple{(:x, :y)}, (z=33, x=1))
 end
 
+@testset "empty children" begin
+  struct EmptyChildren; a; b; end
+  @functor EmptyChildren ()
+  x = EmptyChildren(1, 2)
+  @test children(x) === (;)
```
Comment (CarloLucibello) [inline test/basics.jl:156]:
```suggestion
  @test Functors.children(x) === (;)
```

File: test/basics.jl [hunk 2]
```diff
@@ -149,6 +149,14 @@ end
...
+  @test isleaf(x)
```
Comment (CarloLucibello) [inline test/basics.jl:157]:
```suggestion
  @test Functors.isleaf(x)
```

File: src/Functors.jl [hunk 3]
```diff
@@ -71,6 +71,7 @@ var"@functor"
     Functors.isleaf(x)
 
 Return true if `x` has no [`children`](@ref) according to [`functor`](@ref).
+Corresponding to `isempty(children(x)) == true`.
```
Comment (CarloLucibello) [inline src/Functors.jl:74]:
```suggestion
Corresponds to `isempty(children(x)) == true`.
```

**Comment by CarloLucibello** [comment]
I don't see why we shouldn't do this. I think that `[]`, `()`, `(;)` can be safely considered as leafs

**Comment by ToucheSir** [comment]
It would be pretty breaking, because we use `exclude = isleaf` to decide what to pass to map function callbacks. Code that never expected to encounter an empty (non-numeric) array or dict before is now going to have to deal with them. At a higher level, making two possible paths for particular types based on their contents over just types blows up any chance of us making traversals type stable and introduces additional overhead for users (who will now have to detect and guard against random empty arrays/dicts which they're not interested in. Distinguishing those from parameters would not be fun!)

As I mentioned in https://github.com/FluxML/Functors.jl/issues/52#issuecomment-1327980603, the safer path seems to be limiting and not broadening the scope of `isleaf` so that we have fewer priveleged types.

**Comment by CarloLucibello** [comment]
right, I clearly didn't think it through.

**Pull request closed without merging**